### PR TITLE
feat(0007-backfill): skip unnecessary dictionary lookups

### PR DESCRIPTION
**Pull request by yakkomajuri**
Putting this up as a PoC for @macobo to look over so we can frame our discussion.

This should be able to give us some sort of performance boost in 2 ways:

1. Preventing source table queries that otherwise wouldn't happen: the largest performance boost we can have comes from removing source table queries from the dictionary (i.e. looking up keys that are not cached). This PR will likely remove some of these lookups but probably not _too_ many given it's still likely another row will trigger the query.
2. Preventing dictionary lookups: the marginal benefit of this over each row is tiny (compared to the above), but if we can shave a few microseconds off of each row-column update, over billions of row-columns this would have an impact

### CODE REVIEW

File: posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py [hunk 1]
```diff
@@ -346,34 +346,86 @@ def _run_backfill_mutation(self, query_id):
...
+                    person_properties = if(
+                        person_properties = '',
+                        dictGetStringOrDefault(
+                            '{settings.CLICKHOUSE_DATABASE}.person_dict',
+                            'properties',
+                            tuple(
+                                team_id,
+                                toUUID(dictGet('{settings.CLICKHOUSE_DATABASE}.person_distinct_id2_dict', 'person_id', tuple(team_id, distinct_id)))
+                            ),
+                            toJSONString(map()
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py:363]:
```suggestion
                            toJSONString(map())
```

File: posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py [hunk 2]
```diff
@@ -346,34 +346,86 @@ def _run_backfill_mutation(self, query_id):
...
+                        group2_created_at
+                    ),
+                    group3_created_at = if(
+                        group3_created_at = toDateTime(0),
+                        dictGetDateTime('{settings.CLICKHOUSE_DATABASE}.groups_dict', 'created_at', tuple(team_id, 3, $group_3)),
+                        group3_created_at
+                    ),
+                    group4_created_at = if(
+                        group4_created_at = toDateTime(0),
+                        dictGetDateTime('{settings.CLICKHOUSE_DATABASE}.groups_dict', 'created_at', tuple(team_id, 4, $group_4))
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py:426]:
```suggestion
                        dictGetDateTime('{settings.CLICKHOUSE_DATABASE}.groups_dict', 'created_at', tuple(team_id, 4, $group_4)),
```

**Comment by yakkomajuri** [comment]
Odd that these are failing. Tested that the statements work correctly:

```sql
SELECT if(
    person_created_at = toDateTime(0),
    dictGetDateTime(
        'posthog.person_dict',
        'created_at',
        tuple(
            team_id,
            toUUID(dictGet('posthog.person_distinct_id2_dict', 'person_id', tuple(team_id, distinct_id)))
        )
    ),
    person_created_at
)
FROM sharded_events 
LIMIT 1
```

And also that the `if` doesn't behave weirdly during an update. Still digging...

**Comment by macobo** [comment]
LGTM

Not sure about the test failures - needs a rebase potentially?

**Pull request merged**